tmt/base.py

```diff
@@ -1714,13 +1721,44 @@ def import_plan(self) -> Optional['Plan']:
                 else:
                     tmt.utils.git_clone(plan_id.url, destination, self)
                 if plan_id.ref:
-                    self.run(Command('git', 'checkout', plan_id.ref), cwd=destination)
+                    # attempt to evaluate dynamic reference
+                    try:
+                        # Would it hurt if we simply rewrite plan_id.ref? Better choose safer path.
+                        plan_id._ref_expanded = resolve_dynamic_ref(
+                            workdir=destination,
+                            ref=plan_id.ref,
+                            plan=self,
+                            logger=self._logger
+                            )
+                    except tmt.utils.FileError as error:
+                        raise tmt.utils.DiscoverError(
+                            f"Failed to resolve dynamic ref of '{plan_id.ref}'.") from error
+                    if plan_id._ref_expanded:
+                        self.run(
+                            Command(
+                                'git',
+                                'checkout',
+                                plan_id._ref_expanded),
+                            cwd=destination)
                 if plan_id.path:
                     destination = destination / plan_id.path.unrooted()
                 node = fmf.Tree(str(destination)).find(plan_id.name)
 
             # Use fmf cache for exploring plans (the whole git repo is not needed)
             else:
+                # fmf does not support dynamic ref so when working only with
+                # the fmf cache we won't be able to read and process the dynamic_ref file
+                # pointing us to the right branch.
+                # The whole point of working with cache is to speed up things so it
+                # probably doesn't make sense to clone whole repo.
+                # Therefore, we have decided to lie here use a default branch instead.
+                # An alternative would be either to enforce full clonning or to fail.
```

Comment:
One more idea: The `fmf` cache contains everything which is needed. We could do something like this:

 * fetch dynamic reference using `fmf.Tree.node()` (it is an `fmf` file, right?)
 * apply the dynamic referencing
 * use `fmf.Tree.node()` fetch the plan from the appropriate branch

But this would mean that `resolve_dynamic_ref()` would have to be adjusted to support consuming already fetched dynamic rules. Btw, I just realized, why is it parsing the file manually instead of using `fmf.Tree` methods?

Reply:
> The `fmf` cache contains everything which is needed. We could do

I am afraid I do not understand.

> , why is it parsing the file manually instead of using fmf.Tree methods?

What manual parsing do you mean?

Reply:
> What manual parsing do you mean?

https://github.com/teemtee/tmt/blob/45471b893ac63c1372255d4e0f58d456b7b4b85d/tmt/base.py#L3292-L3299

Instead we could just use:

```python
ref = fmf.Tree.node({"url": "https://github.com/teemtee/tmt/", "path": "/tests/discover/data", "name": "/dynamic-ref"})
```

But the dynamic spec would need to be updated to expect an `fmf` node name:

```yaml
ref: "@/tests/dynamic-ref"
```

Instead of the file path:

```yaml
ref: "@tests/dynamic-ref.fmf"
```

However, this would need some more changes. So, I guess, for now, showing the warning is probably the best way.